idflakies-core/src/main/java/edu/illinois/cs/dt/tools/utility/Tuscan.java

```diff
@@ -0,0 +1,149 @@
+package edu.illinois.cs.dt.tools.utility;
+
+public class Tuscan {
+    public static int[][] generateTuscanPermutations(int arg) {
+        if (arg == 3) {
+            generateThree();
+        } else if (arg == 5) {
+            generateFive();
+        } else {
+            generateTuscanSquare(arg);
+        }
+        return r;
+    }
+    private static int[][] r;
+
+    private static void helper(int[] a, int i) {
+	    System.arraycopy(a, 0, r[i], 0, a.length);
+    }
+
+    private static void generateTuscanSquare(int n) {
+    	int nn = n;
+    	while((n - 1) % 4 == 0 && n != 1 && n != 9){
+             n = (n - 1) / 2 + 1;
+        }
+	    r = new int[nn][];
+	    for (int i = 0; i < nn; i++) {
+	        r[i] = new int[nn + 1];
+	    }
+
+        if (n % 2 == 0) {
+            // https://mathoverflow.net/questions/60856/hamilton-paths-in-k-2n/60859#60859
+            int[] a = new int[n];
+            for (int i = 0; i < n; i += 2) {
+                a[i] = i / 2;
+                a[i + 1] = n - 1 - a[i];
+            }
+            helper(a, 0);
+            for (int j = 1; j < n; j++) {
+                for (int i = 0; i < n; i++) {
+                    a[i] = (a[i] + 1) % n;
+                }
+                helper(a, j);
+            }
+        } else if (n % 4 == 3) {
+            int k = (n - 3) / 4;
+            int[] b = new int[n];
+            for (int i = 0; i < n - 1; i++) {
+                int p = (i == 0) ? 1 :
+                        ((i == k + 1) ? 4 * k + 2 :
+                        ((i == 2 * k + 2) ? 3 :
+                        ((i == 3 * k + 2) ? 4 * k : 2 * k)));
+                int[] a = new int[n];
+                for (int j = 0; j < n; j++) {
+                    a[(j < p) ? n + j - p : j - p] = (j == 0) ? (n - 1) : (i + ((j % 2 == 0) ? (j / 2) : (n - 1 - (j - 1) / 2))) % (n - 1);
+                }
+                b[a[n - 1]] = a[0];
+                helper(a, i);
+            }
+            int[] t = new int[n];
+            t[0] = n - 1;
+            for (int i = 1; i < n; i++) {
+                t[i] = b[t[i - 1]];
+            }
+            helper(t, n - 1);
+        } else if (n == 9) {
+            int[][] t = { 
```

Comment:
Are we sure that in the case of n == 9 that it will also ensure each row has that extra 0 at the end?